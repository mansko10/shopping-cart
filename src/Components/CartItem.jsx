import { useEffect, useState } from "react";

export default function CartItem({
  product,
  updateProductQuantity,
  removeFromCart,
}) {
  const [quantity, setQuantity] = useState(product.quantity);

  useEffect(() => {
    if (quantity !== "") {
      updateProductQuantity(product, quantity);
    }
  }, [quantity]);

  function handleChange(e) {
    setQuantity(e.target.value === "" ? "" : +e.target.value);
  }

  function increaseQuantity() {
    if (quantity === "") {
      setQuantity(1);
    } else {
      setQuantity(quantity + 1);
    }
  }

  function decreaseQuantity() {
    if (quantity === "") {
      setQuantity(1);
    } else if (quantity > 1) {
      setQuantity(quantity - 1);
    }
  }

  return (
    <div className="cart-item">
      <img src={product.image} alt="" />
      <p className="title">{product.title}</p>
      <p className="totalPrice">${+product.totalPrice.toFixed(2)}</p>
      <div className="modify-quantity">
        <button type="button" className="modify" onClick={decreaseQuantity}>
          -
        </button>
        <input type="number" value={quantity} onChange={handleChange} min={1} />
        <button type="button" className="modify" onClick={increaseQuantity}>
          +
        </button>
      </div>
      <button
        type="button"
        className="delete"
        onClick={() => removeFromCart(product)}
      >
        Remove
      </button>
    </div>
  );
}
