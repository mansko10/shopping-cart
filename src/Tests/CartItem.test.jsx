import { describe, expect, it, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CartItem from "../Components/CartItem";

describe("CartItem.jsx", () => {
  const cartProduct = {
    title: "Mens Casual Premium Slim Fit T-Shirts",
    quantity: 2,
    price: 22.3,
    totalPrice: 44.6,
    id: 2,
    image:
      "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
  };

  it("Renders Cartitem.jsx correctly", () => {
    const updateProductQuantity = vi.fn();
    const removeFromCart = vi.fn();

    const { container } = render(
      <CartItem
        product={cartProduct}
        updateProductQuantity={updateProductQuantity}
        removeFromCart={removeFromCart}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it("Should call updateProductQuantity function when '+' is clicked", async () => {
    const user = userEvent.setup();

    const updateProductQuantity = vi.fn();
    const removeFromCart = vi.fn();

    render(
      <CartItem
        product={cartProduct}
        updateProductQuantity={updateProductQuantity}
        removeFromCart={removeFromCart}
      />,
    );

    const increment = screen.getByRole("button", { name: "+" });

    await user.click(increment);

    expect(updateProductQuantity).toHaveBeenCalled();
  });

  it("Should call updateProductQuantity function when '-' button is clicked", async () => {
    const user = userEvent.setup();

    const updateProductQuantity = vi.fn();
    const removeFromCart = vi.fn();

    render(
      <CartItem
        product={cartProduct}
        updateProductQuantity={updateProductQuantity}
        removeFromCart={removeFromCart}
      />,
    );

    const decrement = screen.getByRole("button", { name: "-" });

    await user.click(decrement);

    expect(updateProductQuantity).toHaveBeenCalled();
  });

  it("Should call removeFromCart function when 'remove' button is clicked", async () => {
    const user = userEvent.setup();

    const updateProductQuantity = vi.fn();
    const removeFromCart = vi.fn();

    render(
      <CartItem
        product={cartProduct}
        updateProductQuantity={updateProductQuantity}
        removeFromCart={removeFromCart}
      />,
    );

    const remove = screen.getByRole("button", { name: "Remove" });

    await user.click(remove);

    expect(removeFromCart).toHaveBeenCalled();
  });
});
