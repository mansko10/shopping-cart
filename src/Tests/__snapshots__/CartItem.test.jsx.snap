// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartItem.jsx > Renders Cartitem.jsx correctly 1`] = `
<div>
  <div
    class="cart-item"
  >
    <img
      alt=""
      src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
    />
    <p
      class="title"
    >
      Mens Casual Premium Slim Fit T-Shirts
    </p>
    <p
      class="totalPrice"
    >
      $
      44.6
    </p>
    <div
      class="modify-quantity"
    >
      <button
        class="modify"
        type="button"
      >
        -
      </button>
      <input
        min="1"
        type="number"
        value="2"
      />
      <button
        class="modify"
        type="button"
      >
        +
      </button>
    </div>
    <button
      class="delete"
      type="button"
    >
      Remove
    </button>
  </div>
</div>
`;
